### ISSUE TITLE
[BUG] Bokeh randomly switches backend (webgl <-> canvas) when changing some properties

### ISSUE THREAD

**Issue by stas-sl** [bokeh/bokeh#12022]
### Software versions

Python version      :  3.9.7 (default, Sep 16 2021, 08:50:36)
IPython version     :  7.31.1
Tornado version     :  6.1
Bokeh version       :  2.4.2
BokehJS static path :  /Users/stas/.pyenv/versions/anaconda3-2021.11/lib/python3.9/site-packages/bokeh/server/static
node.js version     :  v17.4.0
npm version         :  8.3.1
Operating system    :  macOS-10.15.7-x86_64-i386-64bit

### Browser name and version

Chrome Version 98.0.4758.102 (Official Build) (x86_64)
Safari Version 14.0.2 (15610.3.7.1.10, 15610)

### Jupyter notebook / Jupyter Lab version

jupyterlab       : 3.2.9

### Expected behavior

Plot backend stays the same as it was specified when creating the figure

### Observed behavior

I created a simple scatter plot and a slider which is linked to glyph size. I used webgl backend because I had a lot of data, but when I moved size slider I observed that after some movements the plot starts to feel slow, then I tried to zoom-in and observed that it starts to use level-of-detail decimation which is applied only in canvas mode (if I understand correctly). Then after more slider movements it switches back to webgl and so on.

### Example code

```Python
import numpy as np

from bokeh.plotting import output_notebook, show, figure
from bokeh.models import Slider
from bokeh.layouts import column

output_notebook()

n = 100
p = figure(output_backend='webgl', width=300, height=300, 
           lod_threshold=50, active_scroll='wheel_zoom')
r = p.scatter(np.random.rand(n), np.random.rand(n), size=10)
s = Slider(title='size', start=1, end=100, value=10)
s.js_link('value', r.glyph, 'size')

show(column(s, p))
```

### Screenshots

https://user-images.githubusercontent.com/4602302/155244660-6cd50188-1a4a-4a19-859a-a931a9b22670.mp4

**Comment by mattpap** [comment]
Seems to be fixed in `branch-3.0`.

**Comment by stas-sl** [comment]
Thanks, good to know, but when are you releasing it? :) Maybe you have ideas how to deal with it in the meantime? It looks quite weird to me - I just don't have any idea why this could happen, shouldn't the backend be initialized just once at creation?

**Comment by ianthomas23** [comment]
I have also confirmed that I can reproduce this using `2.4.2` and that it is fixed in `branch-3.0`.

@stas-sl The webgl backend doesn't work like that. Only certain glyphs have webgl implementations and when you say `output_backend='webgl'`it is not a demand that only webgl will be used but a request to use webgl where available and fall back to canvas if not. We fixed a couple of bugs since 2.4.2 which caused this fall back to be more aggressive than it should be.

Release of 3.0 is being discussed and is not far away, but the timing is determined by the completion of a specific set of functional changes, not a calendar date. In the meantime, you can use `branch-3.0` from the github repo.

Thanks for the high quality issue report by the way.

**Comment by stas-sl** [comment]
Thanks, @ianthomas23, after some debugging I've found that `p.circle` instead of `p.scatter` works without that issue, looks like `ScatterView` has more complicated logic related to webgl than `CircleView`, which contained this bug. So in the meanwhile I'll just use circles, as, actually, I do not need any fancy markers, but looking forward to 3.0 release!
